server/channels/web/oauth.go

```diff
@@ -536,12 +537,31 @@ func signupWithOAuth(c *Context, w http.ResponseWriter, r *http.Request) {
 
 func fullyQualifiedRedirectURL(siteURLPrefix, targetURL string) string {
 	parsed, _ := url.Parse(targetURL)
-	if parsed == nil || parsed.Scheme != "" || parsed.Host != "" {
-		return targetURL
+	prefixParsed, _ := url.Parse(siteURLPrefix)
+
+	// Check if the targetURL is a valid URL and is within the siteURLPrefix
+	if parsed != nil && prefixParsed != nil {
+		sameScheme := parsed.Scheme == prefixParsed.Scheme
+		sameHost := parsed.Host == prefixParsed.Host
+		safePath := strings.HasPrefix(path.Clean(parsed.Path), path.Clean(prefixParsed.Path))
+
+		if sameScheme && sameHost && safePath {
+			return targetURL
+		} else if parsed.Scheme != "" || parsed.Host != "" {
+			return siteURLPrefix
+		}
 	}
 
+	// For relative URLs, normalize and join with siteURLPrefix
 	if targetURL != "" && targetURL[0] != '/' {
 		targetURL = "/" + targetURL
 	}
-	return siteURLPrefix + targetURL
+
+	// Check for path traversal
+	parsed, _ = url.Parse(siteURLPrefix + targetURL)
+	if !strings.HasPrefix(path.Clean(parsed.Path), path.Clean(prefixParsed.Path)) {
```

Comment:
We need to ensure `parsed` is not `nil` or invalid inputs like `http://example.com/%zz` would cause a panic here. Probably worth adding a case in our unit test.